btf/core.go

```diff
@@ -221,14 +230,14 @@ func CORERelocate(relos []*CORERelocation, target *Spec, bo binary.ByteOrder, re
 		group.indices = append(group.indices, i)
 	}
 
+	mergeTarget := mergedSpec(targets)
 	for localType, group := range relosByType {
 		localTypeName := localType.TypeName()
 		if localTypeName == "" {
 			return nil, fmt.Errorf("relocate unnamed or anonymous type %s: %w", localType, ErrNotSupported)
 		}
 
-		targets := target.namedTypes[newEssentialName(localTypeName)]
-		fixups, err := coreCalculateFixups(group.relos, target, targets, bo)
+		fixups, err := coreCalculateFixups(group.relos, &mergeTarget, mergeTarget.NamedTypesIterate(newEssentialName(localTypeName)), bo)
```

Comment:
Merging specs up front is going to be more complicated with lazy copy going in. How about the following pseudo go:

```go
for localType, group := range relosByType {
    var bestFixup
    for _, target := range targets {
        fixup, err := coreCalculateFixup(...)
        
        if fixup > bestFixup {
            bestFixup = fixup
            // other checks
        }
    }
    if bestFixup == nil {
        // generate poison all
    }
}
```

* Instead of merging the specs, we add a loop here which iterates over `targets` and tries to find a valid fixup.
* We factor out the code to do scoring and call if from here in addition to coreCalculateFixups https://github.com/cilium/ebpf/blob/f95957d1669c9d2ee4d6dc34a6dfa25c4130003d/btf/core.go#L276-L294
* We move generating the poison all fixup to after the loop over targets in this function: https://github.com/cilium/ebpf/blob/f95957d1669c9d2ee4d6dc34a6dfa25c4130003d/btf/core.go#L298-L309

Reply:
I'm not actually merging them, just storing pointers to each Spec and iterating through them with wrapper functions. I updated to handle your changes in main. Let me know what you think.

Reply:
Okay, works for me. I'm not a big fan of `mergedSpec`, it seems too special purpose for me. Can you move it to `core.go` so that it's close to where it'll be used?